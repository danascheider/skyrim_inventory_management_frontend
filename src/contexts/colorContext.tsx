import { createContext, ReactElement } from 'react'
import { YELLOW, type ColorScheme } from '../utils/colorSchemes'

const ColorContext = createContext<ColorScheme>(YELLOW)

interface ProviderProps {
  colorScheme: ColorScheme
  children: ReactElement
}

const ColorProvider = ({ colorScheme, children }: ProviderProps) => (
  <ColorContext.Provider value={colorScheme}>{children}</ColorContext.Provider>
)

export { ColorProvider, ColorContext }
