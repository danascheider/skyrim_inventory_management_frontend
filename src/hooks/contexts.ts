import { useContext } from 'react'
import { ColorContext } from '../contexts/colorContext'
import { LoginContext } from '../contexts/loginContext'

const useCustomContext = <T>(cxt: React.Context<T>, msg: string) => {
  const context = useContext(cxt)

  if (!context) throw new Error(msg)

  return context
}

export const useColorScheme = () =>
  useCustomContext(
    ColorContext,
    'useColorScheme must be used within a ColorProvider'
  )

export const useGoogleLogin = () =>
  useCustomContext(
    LoginContext,
    'useGoogleLogin must be used within a LoginProvider'
  )