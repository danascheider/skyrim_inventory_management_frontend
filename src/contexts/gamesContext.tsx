import { createContext, useCallback, useEffect, useState } from 'react'
import { signOutWithGoogle } from '../firebase'
import { type Game } from '../types/games'
import { type ProviderProps } from '../types/contexts'
import { useGoogleLogin, usePageContext } from '../hooks/contexts'
import { ApiError } from '../types/errors'
import { LOADING, DONE, ERROR, type LoadingState } from '../utils/loadingStates'
import { getGames, deleteGame } from '../utils/simApi'

// TODO: Add default values for context provider as follows:
//       const createGame = (game: Game) => {
//                                            postGames(game: Game, token: '')
//                                              .then((status: number, json: string) => /* noop */)
//                                          }
//       This will be useful in Storybook.

const NOT_FOUND_MESSAGE =
  "Oops! We couldn't find the game you're looking for. Please refresh and try again."
const UNEXPECTED_ERROR_MESSAGE =
  "Oops! Something unexpected went wrong. We're sorry! Please try again later."

export interface GamesContextType {
  games: Game[]
  gamesLoadingState: LoadingState
  // createGame: (game: Game) => void
  // updateGame: (gameId: number, attrs: Game) => void
  destroyGame: (gameId: number) => void
}

export const GamesContext = createContext<GamesContextType>({
  games: [],
  gamesLoadingState: LOADING,
  destroyGame: () => {},
})

export const GamesProvider = ({ children }: ProviderProps) => {
  const { user, token, authLoading, requireLogin } = useGoogleLogin()
  const [gamesLoadingState, setGamesLoadingState] = useState(LOADING)
  const [games, setGames] = useState<Game[]>([])
  const { setFlashProps } = usePageContext()

  /*
   *
   * Retrieve all the current user's games from the API
   *
   */

  const fetchGames = useCallback(() => {
    if (user && token) {
      getGames(token)
        .then(({ json }) => {
          if (Array.isArray(json)) {
            setGames(json)
            setGamesLoadingState(DONE)
          }
        })
        .catch((e: ApiError) => {
          if (import.meta.env.DEV)
            console.error(`Error ${e.name}: ${e.message}`)

          if (e.code === 401) {
            signOutWithGoogle()
          } else {
            setGames([])
            setGamesLoadingState(ERROR)
            setFlashProps({
              type: 'error',
              message: e.message,
              hidden: false,
            })
          }
        })
    }
  }, [user, token])

  /*
   *
   * Destroy the requested game and update the `games` array
   *
   */

  const destroyGame = useCallback(
    (gameId: number) => {
      if (user && token) {
        deleteGame(gameId, token)
          .then(({ status }) => {
            if (status === 204) {
              const newGames = games.filter(({ id }) => id !== gameId)
              setGames(newGames)
              setFlashProps({
                hidden: false,
                type: 'success',
                message: 'Success! Your game has been deleted.',
              })
            }
          })
          .catch((e: ApiError) => {
            if (import.meta.env.DEV)
              console.log(`Error ${e.name}: ${e.message}`)

            const message =
              e.code === 404 ? NOT_FOUND_MESSAGE : UNEXPECTED_ERROR_MESSAGE

            setFlashProps({
              hidden: false,
              type: 'error',
              message,
            })
          })
      }
    },
    [user, token, games]
  )

  const value = {
    games,
    gamesLoadingState,
    destroyGame,
  }

  useEffect(() => {
    requireLogin()
  }, [requireLogin])

  useEffect(() => {
    if (authLoading) return

    fetchGames()
  }, [authLoading, fetchGames])

  return <GamesContext.Provider value={value}>{children}</GamesContext.Provider>
}
