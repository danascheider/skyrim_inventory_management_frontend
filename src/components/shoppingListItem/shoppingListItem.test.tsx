import { ReactElement } from 'react'
import { describe, test, expect, vitest, afterEach } from 'vitest'
import { act, fireEvent } from '@testing-library/react'
import { renderAuthenticated } from '../../support/testUtils'
import {
  gamesContextValue,
  shoppingListsContextValue,
} from '../../support/data/contextValues'
import { GREEN } from '../../utils/colorSchemes'
import { PageProvider } from '../../contexts/pageContext'
import { GamesContext } from '../../contexts/gamesContext'
import { ShoppingListsContext } from '../../contexts/shoppingListsContext'
import { ColorProvider } from '../../contexts/colorContext'
import ShoppingListItem from './shoppingListItem'

let listsContextValue = shoppingListsContextValue

const renderInContexts = (ui: ReactElement) =>
  renderAuthenticated(
    <PageProvider>
      <GamesContext.Provider value={gamesContextValue}>
        <ShoppingListsContext.Provider value={listsContextValue}>
          <ColorProvider colorScheme={GREEN}>{ui}</ColorProvider>
        </ShoppingListsContext.Provider>
      </GamesContext.Provider>
    </PageProvider>
  )

describe('ShoppingListItem', () => {
  describe('displaying the list item', () => {
    describe('when the list item is editable', () => {
      test('displays the attributes', () => {
        const wrapper = renderInContexts(
          <ShoppingListItem
            itemId={33}
            listTitle="Clothing"
            description="Silver Necklace"
            quantity={2}
            unitWeight={0.3}
            notes="To enchant"
            editable
          />
        )

        expect(wrapper.getByText('Silver Necklace')).toBeTruthy()
        expect(wrapper.getByText('2')).toBeTruthy()
        expect(wrapper.getByText('0.3')).toBeTruthy()
      })

      test('truncates unit weight values appropriately', () => {
        const wrapper = renderInContexts(
          <ShoppingListItem
            itemId={20}
            listTitle="Clothing"
            description="Silver Necklace"
            quantity={2}
            unitWeight={1.0}
            notes="To enchant"
            editable
          />
        )

        expect(wrapper.getByText('Silver Necklace')).toBeTruthy()
        expect(wrapper.getByText('2')).toBeTruthy()
        expect(wrapper.getByText('1')).toBeTruthy()
        expect(wrapper.queryByText('1.0')).toBeFalsy()
      })

      test('has a destroy icon', () => {
        const wrapper = renderInContexts(
          <ShoppingListItem
            itemId={20}
            listTitle="Clothing"
            description="Silver Necklace"
            quantity={2}
            unitWeight={1.0}
            notes="To enchant"
            editable
          />
        )

        expect(wrapper.getByTestId('destroyShoppingListItem20')).toBeTruthy()
      })

      test('has an edit icon', () => {
        const wrapper = renderInContexts(
          <ShoppingListItem
            itemId={20}
            listTitle="Clothing"
            description="Silver Necklace"
            quantity={2}
            unitWeight={1.0}
            notes="To enchant"
            editable
          />
        )

        expect(wrapper.getByTestId('editShoppingListItem20')).toBeTruthy()
      })

      test('has increment and decrement icons', () => {
        const wrapper = renderInContexts(
          <ShoppingListItem
            itemId={20}
            listTitle="Clothing"
            description="Silver Necklace"
            quantity={2}
            unitWeight={1.0}
            notes="To enchant"
            editable
          />
        )

        expect(wrapper.getByTestId('incrementShoppingListItem20')).toBeTruthy()
        expect(wrapper.getByTestId('decrementShoppingListItem20')).toBeTruthy()
      })

      test('matches snapshot', () => {
        const wrapper = renderInContexts(
          <ShoppingListItem
            itemId={2}
            listTitle="Clothing"
            description="Silver Necklace"
            quantity={2}
            unitWeight={0.3}
            notes="To enchant"
            editable
          />
        )

        expect(wrapper).toMatchSnapshot()
      })
    })

    describe('when the list item is not editable', () => {
      test('displays the attributes', () => {
        const wrapper = renderInContexts(
          <ShoppingListItem
            itemId={33}
            listTitle="All Items"
            description="Silver Necklace"
            quantity={2}
            unitWeight={0.3}
            notes="To enchant"
          />
        )

        expect(wrapper.getByText('Silver Necklace')).toBeTruthy()
        expect(wrapper.getByText('2')).toBeTruthy()
        expect(wrapper.getByText('0.3')).toBeTruthy()
      })

      test('truncates unit weight values appropriately', () => {
        const wrapper = renderInContexts(
          <ShoppingListItem
            itemId={20}
            listTitle="All Items"
            description="Silver Necklace"
            quantity={2}
            unitWeight={1.0}
            notes="To enchant"
          />
        )

        expect(wrapper.getByText('Silver Necklace')).toBeTruthy()
        expect(wrapper.getByText('2')).toBeTruthy()
        expect(wrapper.getByText('1')).toBeTruthy()
        expect(wrapper.queryByText('1.0')).toBeFalsy()
      })

      test('has no destroy icon', () => {
        const wrapper = renderInContexts(
          <ShoppingListItem
            itemId={20}
            listTitle="All Items"
            description="Silver Necklace"
            quantity={2}
            unitWeight={1.0}
            notes="To enchant"
          />
        )

        expect(wrapper.queryByTestId('destroyShoppingListItem20')).toBeFalsy()
      })

      test('has no edit icon', () => {
        const wrapper = renderInContexts(
          <ShoppingListItem
            itemId={20}
            listTitle="All Items"
            description="Silver Necklace"
            quantity={2}
            unitWeight={1.0}
            notes="To enchant"
          />
        )

        expect(wrapper.queryByTestId('editShoppingListItem20')).toBeFalsy()
      })

      test('has no increment or decrement icons', () => {
        const wrapper = renderInContexts(
          <ShoppingListItem
            itemId={20}
            listTitle="All Items"
            description="Silver Necklace"
            quantity={2}
            unitWeight={1.0}
            notes="To enchant"
          />
        )

        expect(wrapper.queryByTestId('incrementShoppingListItem20')).toBeFalsy()
        expect(wrapper.queryByTestId('decrementShoppingListItem20')).toBeFalsy()
      })

      test('matches snapshot', () => {
        const wrapper = renderInContexts(
          <ShoppingListItem
            itemId={2}
            listTitle="All Items"
            description="Silver Necklace"
            quantity={2}
            unitWeight={0.3}
            notes="To enchant"
          />
        )

        expect(wrapper).toMatchSnapshot()
      })
    })
  })

  describe('destroying the list item', () => {
    describe('when the user confirms deletion', () => {
      const ogConfirm = window.confirm

      afterEach(() => {
        listsContextValue = shoppingListsContextValue
        window.confirm = ogConfirm
      })

      test("calls the context's destroyShoppingListItem function with its itemId", () => {
        const destroyShoppingListItem = vitest.fn()
        listsContextValue = {
          ...shoppingListsContextValue,
          destroyShoppingListItem,
        }

        const wrapper = renderInContexts(
          <ShoppingListItem
            itemId={33}
            listTitle="Clothing"
            description="Silver Necklace"
            quantity={2}
            unitWeight={0.3}
            notes="To enchant"
            editable
          />
        )

        window.confirm = vitest.fn().mockImplementation(() => true)

        const destroyIcon = wrapper.getByTestId('destroyShoppingListItem33')

        act(() => fireEvent.click(destroyIcon))

        expect(window.confirm).toHaveBeenCalled
        expect(destroyShoppingListItem).toHaveBeenCalledWith(33)
      })
    })

    describe('when the user cancels deletion', () => {
      const ogConfirm = window.confirm

      afterEach(() => {
        listsContextValue = shoppingListsContextValue
        window.confirm = ogConfirm
      })

      test("doesn't call the destroyShoppingListItem function", () => {
        const destroyShoppingListItem = vitest.fn()
        listsContextValue = {
          ...shoppingListsContextValue,
          destroyShoppingListItem,
        }

        const wrapper = renderInContexts(
          <ShoppingListItem
            itemId={33}
            listTitle="Clothing"
            description="Silver Necklace"
            quantity={2}
            unitWeight={0.3}
            notes="To enchant"
            editable
          />
        )

        window.confirm = vitest.fn().mockImplementation(() => false)

        const destroyIcon = wrapper.getByTestId('destroyShoppingListItem33')

        act(() => fireEvent.click(destroyIcon))

        expect(window.confirm).toHaveBeenCalledOnce()
        expect(destroyShoppingListItem).not.toHaveBeenCalled()
      })
    })
  })

  describe('incrementing the list item quantity', () => {
    afterEach(() => {
      listsContextValue = shoppingListsContextValue
    })

    test('increments the quantity at the API', () => {
      const updateShoppingListItem = vitest.fn()
      listsContextValue = { ...listsContextValue, updateShoppingListItem }

      const wrapper = renderInContexts(
        <ShoppingListItem
          itemId={33}
          listTitle="Clothing"
          description="Silver Necklace"
          quantity={2}
          unitWeight={0.3}
          notes="To enchant"
          editable
        />
      )

      const incrementIcon = wrapper.getByTestId('incrementShoppingListItem33')

      act(() => {
        fireEvent.click(incrementIcon)
      })

      expect(updateShoppingListItem).toHaveBeenCalledWith(
        33,
        { quantity: 3 },
        expect.any(Function),
        expect.any(Function)
      )
    })
  })

  describe('decrementing the list item quantity', () => {
    afterEach(() => {
      listsContextValue = shoppingListsContextValue
    })

    describe('when the quantity is greater than 1', () => {
      test('decrements the quantity at the API', () => {
        const updateShoppingListItem = vitest.fn()
        listsContextValue = { ...listsContextValue, updateShoppingListItem }

        const wrapper = renderInContexts(
          <ShoppingListItem
            itemId={33}
            listTitle="Clothing"
            description="Silver Necklace"
            quantity={2}
            unitWeight={0.3}
            notes="To enchant"
            editable
          />
        )

        const decrementIcon = wrapper.getByTestId('decrementShoppingListItem33')

        act(() => fireEvent.click(decrementIcon))

        expect(updateShoppingListItem).toHaveBeenCalledWith(
          33,
          { quantity: 1 },
          expect.any(Function),
          expect.any(Function)
        )
      })
    })

    describe('when the quantity is 1', () => {
      const ogConfirm = window.confirm

      afterEach(() => {
        listsContextValue = shoppingListsContextValue
        window.confirm = ogConfirm
      })

      test('prompts the user to destroy the item', () => {
        const wrapper = renderInContexts(
          <ShoppingListItem
            itemId={33}
            listTitle="Clothing"
            description="Silver Necklace"
            quantity={1}
            unitWeight={0.3}
            notes="To enchant"
            editable
          />
        )

        window.confirm = vitest.fn().mockImplementation(() => false)

        const decrementIcon = wrapper.getByTestId('decrementShoppingListItem33')

        act(() => fireEvent.click(decrementIcon))

        expect(window.confirm).toHaveBeenCalledOnce()
      })

      test('destroys the item if the user confirms', () => {
        const destroyShoppingListItem = vitest.fn()
        listsContextValue = {
          ...shoppingListsContextValue,
          destroyShoppingListItem,
        }

        const wrapper = renderInContexts(
          <ShoppingListItem
            itemId={33}
            listTitle="Clothing"
            description="Silver Necklace"
            quantity={1}
            unitWeight={0.3}
            notes="To enchant"
            editable
          />
        )

        window.confirm = vitest.fn().mockImplementation(() => true)

        const decrementIcon = wrapper.getByTestId('decrementShoppingListItem33')

        act(() => fireEvent.click(decrementIcon))

        expect(window.confirm).toHaveBeenCalledOnce()
        expect(destroyShoppingListItem).toHaveBeenCalledWith(
          33,
          null,
          expect.any(Function)
        )
      })
    })
  })
})
