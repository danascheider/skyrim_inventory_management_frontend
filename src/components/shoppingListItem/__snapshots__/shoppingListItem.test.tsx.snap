// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ShoppingListItem > displaying the list item > when the list item is editable > matches snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="_root_fccc8d"
        style="--main-color: #32b54e; --title-text-color: #fff; --border-color: #00821c; --body-background-color: #ccecd3; --body-text-color: #00410e; --hover-color: #19ac38;"
      >
        <div
          aria-controls="shoppingListItem2Details"
          aria-expanded="false"
          class="_trigger_fccc8d"
        >
          <span
            class="_descriptionContainer_fccc8d _descriptionContainerEditable_fccc8d"
          >
            <span
              class="_editIcons_fccc8d"
            >
              <button
                class="_icon_fccc8d"
                data-testid="destroyShoppingListItem2"
              >
                <svg
                  aria-hidden="true"
                  class="svg-inline--fa fa-xmark _fa_fccc8d"
                  data-icon="xmark"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  viewBox="0 0 384 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M342.6 150.6c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L192 210.7 86.6 105.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3L146.7 256 41.4 361.4c-12.5 12.5-12.5 32.8 0 45.3s32.8 12.5 45.3 0L192 301.3 297.4 406.6c12.5 12.5 32.8 12.5 45.3 0s12.5-32.8 0-45.3L237.3 256 342.6 150.6z"
                    fill="currentColor"
                  />
                </svg>
              </button>
              <button
                class="_icon_fccc8d"
                data-testid="editShoppingListItem2"
              >
                <svg
                  aria-hidden="true"
                  class="svg-inline--fa fa-pen-to-square _fa_fccc8d"
                  data-icon="pen-to-square"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  viewBox="0 0 512 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M471.6 21.7c-21.9-21.9-57.3-21.9-79.2 0L362.3 51.7l97.9 97.9 30.1-30.1c21.9-21.9 21.9-57.3 0-79.2L471.6 21.7zm-299.2 220c-6.1 6.1-10.8 13.6-13.5 21.9l-29.6 88.8c-2.9 8.6-.6 18.1 5.8 24.6s15.9 8.7 24.6 5.8l88.8-29.6c8.2-2.7 15.7-7.4 21.9-13.5L437.7 172.3 339.7 74.3 172.4 241.7zM96 64C43 64 0 107 0 160V416c0 53 43 96 96 96H352c53 0 96-43 96-96V320c0-17.7-14.3-32-32-32s-32 14.3-32 32v96c0 17.7-14.3 32-32 32H96c-17.7 0-32-14.3-32-32V160c0-17.7 14.3-32 32-32h96c17.7 0 32-14.3 32-32s-14.3-32-32-32H96z"
                    fill="currentColor"
                  />
                </svg>
              </button>
            </span>
            <h3
              class="_description_fccc8d"
            >
              Silver Necklace
            </h3>
          </span>
          <span
            class="_quantityEditable_fccc8d"
          >
            <button
              class="_icon_fccc8d"
              data-testid="incrementShoppingListItem2"
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-chevron-up _fa_fccc8d"
                data-icon="chevron-up"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 512 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M233.4 105.4c12.5-12.5 32.8-12.5 45.3 0l192 192c12.5 12.5 12.5 32.8 0 45.3s-32.8 12.5-45.3 0L256 173.3 86.6 342.6c-12.5 12.5-32.8 12.5-45.3 0s-12.5-32.8 0-45.3l192-192z"
                  fill="currentColor"
                />
              </svg>
            </button>
            <span
              class="_quantityContent_fccc8d"
            >
              2
            </span>
            <button
              class="_icon_fccc8d"
              data-testid="decrementShoppingListItem2"
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-chevron-down _fa_fccc8d"
                data-icon="chevron-down"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 512 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M233.4 406.6c12.5 12.5 32.8 12.5 45.3 0l192-192c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L256 338.7 86.6 169.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3l192 192z"
                  fill="currentColor"
                />
              </svg>
            </button>
          </span>
        </div>
        <div
          aria-hidden="true"
          class="rah-static rah-static--height-zero "
          id="shoppingListItem2Details"
          style="height: 0px; overflow: hidden;"
        >
          <div
            style="display: none;"
          >
            <div
              class="_details_fccc8d"
            >
              <h4
                class="_label_fccc8d"
              >
                Unit Weight:
              </h4>
              <p
                class="_value_fccc8d"
              >
                0.3
              </p>
              <h4
                class="_label_fccc8d"
              >
                Notes:
              </h4>
              <p
                class="_value_fccc8d"
              >
                To enchant
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="_root_fccc8d"
      style="--main-color: #32b54e; --title-text-color: #fff; --border-color: #00821c; --body-background-color: #ccecd3; --body-text-color: #00410e; --hover-color: #19ac38;"
    >
      <div
        aria-controls="shoppingListItem2Details"
        aria-expanded="false"
        class="_trigger_fccc8d"
      >
        <span
          class="_descriptionContainer_fccc8d _descriptionContainerEditable_fccc8d"
        >
          <span
            class="_editIcons_fccc8d"
          >
            <button
              class="_icon_fccc8d"
              data-testid="destroyShoppingListItem2"
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-xmark _fa_fccc8d"
                data-icon="xmark"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 384 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M342.6 150.6c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L192 210.7 86.6 105.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3L146.7 256 41.4 361.4c-12.5 12.5-12.5 32.8 0 45.3s32.8 12.5 45.3 0L192 301.3 297.4 406.6c12.5 12.5 32.8 12.5 45.3 0s12.5-32.8 0-45.3L237.3 256 342.6 150.6z"
                  fill="currentColor"
                />
              </svg>
            </button>
            <button
              class="_icon_fccc8d"
              data-testid="editShoppingListItem2"
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-pen-to-square _fa_fccc8d"
                data-icon="pen-to-square"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 512 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M471.6 21.7c-21.9-21.9-57.3-21.9-79.2 0L362.3 51.7l97.9 97.9 30.1-30.1c21.9-21.9 21.9-57.3 0-79.2L471.6 21.7zm-299.2 220c-6.1 6.1-10.8 13.6-13.5 21.9l-29.6 88.8c-2.9 8.6-.6 18.1 5.8 24.6s15.9 8.7 24.6 5.8l88.8-29.6c8.2-2.7 15.7-7.4 21.9-13.5L437.7 172.3 339.7 74.3 172.4 241.7zM96 64C43 64 0 107 0 160V416c0 53 43 96 96 96H352c53 0 96-43 96-96V320c0-17.7-14.3-32-32-32s-32 14.3-32 32v96c0 17.7-14.3 32-32 32H96c-17.7 0-32-14.3-32-32V160c0-17.7 14.3-32 32-32h96c17.7 0 32-14.3 32-32s-14.3-32-32-32H96z"
                  fill="currentColor"
                />
              </svg>
            </button>
          </span>
          <h3
            class="_description_fccc8d"
          >
            Silver Necklace
          </h3>
        </span>
        <span
          class="_quantityEditable_fccc8d"
        >
          <button
            class="_icon_fccc8d"
            data-testid="incrementShoppingListItem2"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-chevron-up _fa_fccc8d"
              data-icon="chevron-up"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 512 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M233.4 105.4c12.5-12.5 32.8-12.5 45.3 0l192 192c12.5 12.5 12.5 32.8 0 45.3s-32.8 12.5-45.3 0L256 173.3 86.6 342.6c-12.5 12.5-32.8 12.5-45.3 0s-12.5-32.8 0-45.3l192-192z"
                fill="currentColor"
              />
            </svg>
          </button>
          <span
            class="_quantityContent_fccc8d"
          >
            2
          </span>
          <button
            class="_icon_fccc8d"
            data-testid="decrementShoppingListItem2"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-chevron-down _fa_fccc8d"
              data-icon="chevron-down"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 512 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M233.4 406.6c12.5 12.5 32.8 12.5 45.3 0l192-192c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L256 338.7 86.6 169.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3l192 192z"
                fill="currentColor"
              />
            </svg>
          </button>
        </span>
      </div>
      <div
        aria-hidden="true"
        class="rah-static rah-static--height-zero "
        id="shoppingListItem2Details"
        style="height: 0px; overflow: hidden;"
      >
        <div
          style="display: none;"
        >
          <div
            class="_details_fccc8d"
          >
            <h4
              class="_label_fccc8d"
            >
              Unit Weight:
            </h4>
            <p
              class="_value_fccc8d"
            >
              0.3
            </p>
            <h4
              class="_label_fccc8d"
            >
              Notes:
            </h4>
            <p
              class="_value_fccc8d"
            >
              To enchant
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`ShoppingListItem > displaying the list item > when the list item is not editable > matches snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="_root_fccc8d"
        style="--main-color: #32b54e; --title-text-color: #fff; --border-color: #00821c; --body-background-color: #ccecd3; --body-text-color: #00410e; --hover-color: #19ac38;"
      >
        <div
          aria-controls="shoppingListItem2Details"
          aria-expanded="false"
          class="_trigger_fccc8d"
        >
          <span
            class="_descriptionContainer_fccc8d"
          >
            <h3
              class="_description_fccc8d"
            >
              Silver Necklace
            </h3>
          </span>
          <span
            class="_quantity_fccc8d"
          >
            <span
              class="_quantityContent_fccc8d"
            >
              2
            </span>
          </span>
        </div>
        <div
          aria-hidden="true"
          class="rah-static rah-static--height-zero "
          id="shoppingListItem2Details"
          style="height: 0px; overflow: hidden;"
        >
          <div
            style="display: none;"
          >
            <div
              class="_details_fccc8d"
            >
              <h4
                class="_label_fccc8d"
              >
                Unit Weight:
              </h4>
              <p
                class="_value_fccc8d"
              >
                0.3
              </p>
              <h4
                class="_label_fccc8d"
              >
                Notes:
              </h4>
              <p
                class="_value_fccc8d"
              >
                To enchant
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="_root_fccc8d"
      style="--main-color: #32b54e; --title-text-color: #fff; --border-color: #00821c; --body-background-color: #ccecd3; --body-text-color: #00410e; --hover-color: #19ac38;"
    >
      <div
        aria-controls="shoppingListItem2Details"
        aria-expanded="false"
        class="_trigger_fccc8d"
      >
        <span
          class="_descriptionContainer_fccc8d"
        >
          <h3
            class="_description_fccc8d"
          >
            Silver Necklace
          </h3>
        </span>
        <span
          class="_quantity_fccc8d"
        >
          <span
            class="_quantityContent_fccc8d"
          >
            2
          </span>
        </span>
      </div>
      <div
        aria-hidden="true"
        class="rah-static rah-static--height-zero "
        id="shoppingListItem2Details"
        style="height: 0px; overflow: hidden;"
      >
        <div
          style="display: none;"
        >
          <div
            class="_details_fccc8d"
          >
            <h4
              class="_label_fccc8d"
            >
              Unit Weight:
            </h4>
            <p
              class="_value_fccc8d"
            >
              0.3
            </p>
            <h4
              class="_label_fccc8d"
            >
              Notes:
            </h4>
            <p
              class="_value_fccc8d"
            >
              To enchant
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
